src/robot/Robot.cpp

```diff
@@ -1,21 +1,454 @@
+void Robot::cancelAllTrajectories(
+    bool incluldeSubrobots,
+    bool includeParents,
+    const std::vector<std::string> excludedSubrobots)
+{
+  // Cancel this trajectory
+  if (mTrajectoryExecutor)
+  {
+    mTrajectoryExecutor->cancel();
+  }
+
+  // Cancel parents' trajectories (if requested)
+  if (includeParents && mRootRobot)
+  {
+    mRootRobot->cancelAllTrajectories(false, true);
+  }
+
+  // Cancel children's trajectories
+  if (incluldeSubrobots)
+  {
+    for (const auto& subrobot : mSubRobots)
+    {
+      if (!std::count(
+              excludedSubrobots.begin(),
+              excludedSubrobots.end(),
+              subrobot.first))
+      {
+        subrobot.second->cancelAllTrajectories(true, false);
+      }
+    }
+  }
+}
+
+//==============================================================================
+void Robot::step(const std::chrono::system_clock::time_point& timepoint)
+{
+  // Lock only if root robot
+  std::unique_ptr<std::lock_guard<std::mutex>> lock;
+  if (!mRootRobot)
+  {
+    lock = std::make_unique<std::lock_guard<std::mutex>>(
+        mMetaSkeleton->getBodyNode(0)->getSkeleton()->getMutex());
+  }
+
+  if (mTrajectoryExecutor)
+  {
+    mTrajectoryExecutor->step(timepoint);
+  }
+
+  for (const auto& subrobot : mSubRobots)
+  {
+    subrobot.second->step(timepoint);
+  }
+}
+
+// ==============================================================================
+constraint::dart::CollisionFreePtr Robot::getSelfCollisionConstraint() const
+{
+  // Only use root robot's self-collision constraint
+  if (mRootRobot)
+  {
+    return mRootRobot->getSelfCollisionConstraint();
+  }
+
+  // Add collision option with self-collision filter
+  auto collisionOption
+      = dart::collision::CollisionOption(false, 1, mSelfCollisionFilter);
+
+  // Create the constraint and return.
+  auto collisionFreeConstraint
+      = std::make_shared<constraint::dart::CollisionFree>(
+          mStateSpace,
+          mMetaSkeleton->cloneMetaSkeleton(),
+          mCollisionDetector,
+          collisionOption);
+  collisionFreeConstraint->addSelfCheck(
+      mCollisionDetector->createCollisionGroupAsSharedPtr(mMetaSkeleton.get()));
+  return collisionFreeConstraint;
+}
+
+// ==============================================================================
+constraint::TestablePtr Robot::combineCollisionConstraint(
+    const constraint::dart::CollisionFreePtr& collisionFree) const
+{
+  using constraint::TestableIntersection;
+
+  // Only use root robot's self-collision constraint
+  if (mRootRobot)
+  {
+    return mRootRobot->combineCollisionConstraint(collisionFree);
+  }
+
+  auto selfCollisionFree = getSelfCollisionConstraint();
+
+  if (!collisionFree)
+    return selfCollisionFree;
+
+  // Make testable constraints for collision check
+  std::vector<aikido::constraint::ConstTestablePtr> constraints;
+  constraints.reserve(2);
+  constraints.emplace_back(selfCollisionFree);
+  if (collisionFree)
```

Comment:
Small nit here -- shouldn't this by default be true because you already return if it is not true? I'm not sure if this adds functionality or readability by being here